Speed up ExtractIlluminaBarcodes by not turning every barcode read in…

tfenne opened the pull request:
…to a String.

There's probably more we could do, and cleaning up this code to make it re-usable will help make that visible.  Still, this cut ~25% off the runtime for me.

jacarey commented on src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:
```diff
@@ -598,6 +596,51 @@ public void merge(final BarcodeMetric metric) {
+            // Pre-compute the hash-code
+            int h = 0;
+            for (final byte[] bs : this.bytes) {
+                for (int i=0; i<bs.length; ++i) {
```
```suggestion
                for (final byte b : bs) {
```

jacarey commented on src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:
```diff
@@ -808,32 +854,22 @@ private static boolean ensureLookupMinimumValue(final byte[][] qualityScores, fi
+                                             final int minimumBaseQuality) {
+            final boolean canUseLookupTable = areAllQualitiesAboveMinimum(qualityScores, minimumBaseQuality);
+            final ByteString barcodesAsString = new ByteString(readSubsequences);
+            BarcodeMatch match = null;
+
+            if (canUseLookupTable) match = barcodeLookupMap.get(barcodesAsString);
```
```suggestion
BarcodeMatch match = canUseLookupTable ? barcodeLookupMap.get(barcodesAsString) : null;
```

jacarey commented on src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:
```diff
@@ -218,7 +213,9 @@
     private final Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
-    private final ConcurrentHashMap<String, PerTileBarcodeExtractor.BarcodeMatch> barcodeLookupMap = new ConcurrentHashMap<>();
+
+    private final static int MAX_LOOKUP_SIZE = 100000;
```
I'm curious to see if the timings I go for the MiSeq run make a bigger difference with a novaseq worth of data. Perhaps this should be a tweakable parameter?

```
390625 (5^8) = Runtime: 19.324 s
128k          = Runtime: 19.435 s
106496        = Runtime: 19.048 s
100k          = Runtime: 19.241 s
64k           = Runtime: 20.215 s
32k           = Runtime: 19.65 s
16k           = Runtime: 19.55 s
8k            = Runtime: 19.305 s
4k            = Runtime: 18.516 s
1             = Runtime: 29.713 s
```

tfenne replied:
I tried a little bit with the MiSeq run and it seemed to me that expanding the cache further slowed things down.  My 2c is that the right solution (yet to be proven):
1. Extract/remove any reference to Metric objects from the _matching_ methods so that a) they're cleaner and b) we can pre-fill the cache without having to worry about messing up metrics
2. Pre-fill the cache with all perfect matches, and all no-calls
3. Change the conditional for entry into the cache to be based on whether the barcode read matched a known barcode and _not_ limited by number.

My thought being that there is a lot of value caching things that we see frequently, and little value caching things that we are unlikely to see frequently.

jacarey replied:
The latest code includes the above changes.

jacarey reviewed (approved):
LGTM. Do you want me to pick this branch up for the additional refactor or should we merge this down and start a new branch for that?

tfenne commented:
Yes please on picking up this branch and starting your refactor from here.  Let's not merge this into the other performance branch, but try and keep the changes for ExtractIlluminaBarcodes to a single separate PR.

jacarey commented:
@tfenne Assigned to you since I can't request you as a reviewer on a PR that you created.

jacarey commented on src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:
```diff
@@ -406,13 +440,27 @@ private File getBarcodeFile(final int tile) {
                 }
                 barcodes.add(barcode);
-                final BarcodeMetric metric = new BarcodeMetric(null, null, barcode, new String[]{barcode});
+                int barcodeNum = 0;
```
If a BARCODE was specified and not a BARCODE_FILE the metric map was constructed differently. I normalized this.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class BarcodeExtractor {
```
Class level javadoc please.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+    final BarcodeMetric noMatch;
+    final int maxNoCalls, maxMismatches, minMismatchDelta, minimumBaseQuality;
+    final DistanceMetric distanceMode;
+    final ConcurrentHashMap<ByteString, BarcodeMatch> barcodeLookupMap;
```
Can these all be private?

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+
+    public BarcodeExtractor(final Map<String, BarcodeMetric> barcodeToMetrics,
+                            final ConcurrentHashMap<ByteString, BarcodeMatch> barcodeLookupMap,
+                            Set<ByteString> barcodeByteStrings, final BarcodeMetric noMatchMetric,
```
```suggestion
                            Set<ByteString> barcodeByteStrings,
                            final BarcodeMetric noMatchMetric,
```

tfenne replied:
Do you need `barcodeByteStrings`? Or could you manufacture it in the constructor from the keys in `barcodeToMetrics`?  It would be nice to keep the use of `ByteString` hidden inside `BarcodeExtractor` if possible and not expose other classes to it.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+        }
+
+        return match;
+    }
```
```suggestion
        final boolean canUseLookupTable = areAllQualitiesAboveMinimum(qualityScores, minimumBaseQuality);
        if (canUseLookupTable) {
            final ByteString barcodesAsString = new ByteString(readSubsequences);
            BarcodeMatch match = barcodeLookupMap.get(barcodesAsString);
            if (match == null) match = calculateBarcodeMatch(readSubsequences, qualityScores);
            if (match.isMatched()) barcodeLookupMap.put(barcodesAsString, match);
            return match;
        }
        else {
            return calculateBarcodeMatch(readSubsequences, qualityScores);
        }
    }
```

To avoid making the ByteString if we can't use it, and tidy up a bit.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+                }
+            }
+        }
+
```
I guess it would be a functional change, but I think it would be nice to short-circuit here when `numNoCalls > maxNoCalls` instead of wasting cycles doing all the matching.  The only difference would be in the `s_x_barcode.txt` files that are output, not in the BAMs/fastqs.

tfenne replied:
We could also add a flag to `DistanceMetric` to ask if it considers low-quality bases a mismatch.  And then check up front for the sum of no-calls and low-qual bases, and avoid doing all the matching if the read is too shitty to possibly match.  If we really cared we could even add a flag to `BarcodeExtractor` to toggle this behavior on for demux-on-the-fly and off for "classic".

jacarey replied:
> I guess it would be a functional change, but I think it would be nice to short-circuit here when `numNoCalls > maxNoCalls` instead of wasting cycles doing all the matching. The only difference would be in the `s_x_barcode.txt` files that are output, not in the BAMs/fastqs.

We could do this for inline matching only and not affect the barcode.txt outputs.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+        } else {
+            // If we have something that's not a "match" but matches one barcode
+            // slightly, we output that matching barcode in lower case
+            if (numNoCalls + numMismatchesInBestBarcode < totalBarcodeReadBases && bestBarcode != null) {
+                match.barcode = bestBarcode.toLowerCase();
```
Similarly if we had a flag for `isOnTheFly` we could avoid this lower-casing of the "closest" non-match since it'll never be used.

jacarey replied:
This changes the match.mismatches though which will in turn change the metric no?

Actually, I think the only changes the mismatches for the returned match, but not the final updated metrics.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    static void updateMetrics(final BarcodeMatch match, final boolean passingFilter,
+                                      final Map<String, BarcodeMetric> metrics, final BarcodeMetric noMatchBarcodeMetric) {
```
```suggestion
    static void updateMetrics(final BarcodeMatch match,
                              final boolean passingFilter,
                              final Map<String, BarcodeMetric> metrics,
                              final BarcodeMetric noMatchBarcodeMetric) {
```

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+        boolean matched;
+        String barcode;
+        int mismatches;
+        int mismatchesToSecondBest;
```
Can these all be private?

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,248 @@
+    /**
+     * Class to give a byte[][] a hashcode and equals without copying the whole contents into a String.
+     */
+    static final class ByteString {
```
Would be nice for this to be `private` within `BarcodeExtractor`

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -284,11 +285,12 @@ private void checkCompletedWork() throws InterruptedException {
     protected static Set<IlluminaDataType> getDataTypesFromReadStructure(final ReadStructure readStructure,
-                                                                         final boolean demultiplex) {
-        if (!readStructure.hasSampleBarcode() || !demultiplex) {
+                                                                         final boolean demultiplex, File barcodesDir) {
```
```suggestion
                                                                         final boolean demultiplex,
                                                                         File barcodesDir) {
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -339,6 +341,30 @@ protected void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+    protected String maybeDemultiplex(ClusterData cluster) {
+        String barcode = null;
+        if (demultiplex) {
+            // If there is no barcode parsed from barcode file try on the fly demux.
```
I think this is out of date and should be something like `// if a barcode extractor was provided for on-the-fly demux us it`

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -97,18 +95,21 @@ public BasecallsConverter(
             final boolean includeNonPfReads,
-            final int numWriteThreads
+            final int numWriteThreads,
+            final BarcodeExtractor barcodeExtractor
```
These both need adding to the `@param` list above

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+    public File BASECALLS_DIR;
+
+    @Argument(doc = "Where to write _barcode.txt files.  By default, these are written to BASECALLS_DIR.", optional = true)
+    public File OUTPUT_DIR;
```
Is this one not still specific to ExtractIlluminaBarcodes?

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+    public File METRICS_FILE;
+
+    @Argument
+    public Boolean BARCODE_EXTRACT = false;
```
What's this?

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+    public Boolean BARCODE_EXTRACT = false;
+
+    @Argument(optional = true)
+    public File INPUT_PARAMS_FILE;
```
Ditto - without `doc=""` I don't know what this is.

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+
+    protected IlluminaDataProviderFactory factory;
+    protected final Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
+    protected final static int MAX_LOOKUP_SIZE = 4096;
```
This is really the "initial size" now right?  ConcurrentHashMap is not bounded.

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+            }
+        }
+
+        BarcodeMetric noMatchMetric = new BarcodeMetric(null, null, IlluminaUtil.barcodeSeqsToString(noMatchBarcode), noMatchBarcode);
```
final?

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+
+        for (final BarcodeMetric metric : barcodeToMetrics.values()) {
+            this.barcodeByteStrings.add(new BarcodeExtractor.ByteString(metric.barcodeBytes));
+        }
```
Lets move this into BarcodeExtractor.

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+                MINIMUM_BASE_QUALITY,
+                DISTANCE_MODE);
+
+        // Prepopulate the lookup map with all perfect barcodes
+        for(BarcodeMetric metric : barcodeToMetrics.values()){
+            BarcodeExtractor.BarcodeMatch match = barcodeExtractor.calculateBarcodeMatch(metric.barcodeBytes,
+                    perfectScores);
+            barcodeLookupMap.put(new BarcodeExtractor.ByteString(metric.barcodeBytes), match);
+        }
```
Could this also be done inside the extractor?  Do we need to expose the cache at all even?

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        // Prepopulate all no call barcode match
+        BarcodeExtractor.BarcodeMatch noCallMatch = barcodeExtractor.calculateBarcodeMatch(noMatchMetric.barcodeBytes,
+                perfectScores);
+
+        barcodeLookupMap.put(new BarcodeExtractor.ByteString(noMatchMetric.barcodeBytes), noCallMatch);
```
Given the requirement internally for all bases to be above the score threshold to use the cache, it's not clear to me if this will really get used.

jacarey replied:
Should we allow using the cache when ALL bases are below the quality threshold ? Or just remove the all no call pre-populated value?

tfenne commented on src/main/java/picard/illumina/BarcodeMetric.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Creates a copy of metric initialized with only non-accumulated and non-calculated values set
+     */
+    public static BarcodeMetric copy(final BarcodeMetric metric) {
```
Why not just make this non-static so you can call it  directly on instances instead of passing in an instance?

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -60,7 +55,7 @@
-public class IlluminaBasecallsToFastq extends CommandLineProgram {
+public class IlluminaBasecallsToFastq extends ExtractBarcodesProgram {
     static final String USAGE_SUMMARY = "Generate FASTQ file(s) from Illumina basecall read data.  ";
     static final String USAGE_DETAILS = "<p>This tool generates FASTQ files from data in an Illumina BaseCalls output directory.  " +
```
Should probably update the details to discuss on-the-fly demux and which parameters affect it.

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -104,14 +99,8 @@
-
-    @Argument(doc = "Lane number. ", shortName = StandardOptionDefinitions.LANE_SHORT_NAME)
-    public Integer LANE;
+    public File BARCODES_DIR = BASECALLS_DIR;
```
adding ` = BASECALLS_DIR` here won't have any effect because at contruction time `BASECALLS_DIR` will be null also.

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -214,6 +192,7 @@ protected int doWork() {
 
     @Override
     protected String[] customCommandLineValidation() {
+        INPUT_PARAMS_FILE = MULTIPLEX_PARAMS;
```
Should this be guarded by `if (INPUT_PARAMS_FILE = null)`?

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -281,6 +256,12 @@ private void initialize() {
                 .withIgnoreUnexpectedBarcodes(IGNORE_UNEXPECTED_BARCODES)
                 .withBclQualityEvaluationStrategy(bclQualityEvaluationStrategy);
 
+        if(BARCODE_EXTRACT && demultiplex){
```
I think probably `BARCODE_EXTRACT` should come down into this class (because it doesn't make sense for ExtractIlluminaBarcodes` and be renamed to something like `MATCH_BARCODES_ON_THE_FLY` (better names welcome).

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -301,6 +287,12 @@ private void initialize() {
                 .withIgnoreUnexpectedBarcodes(IGNORE_UNEXPECTED_BARCODES)
                 .withBclQualityEvaluationStrategy(bclQualityEvaluationStrategy);
 
+        if(BARCODE_EXTRACT && demultiplex){
```
Ditto.

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class SampleInputParameters {
```
Need javadoc

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,100 @@
+    public static final String LIBRARY_NAME_COLUMN = "library_name";
+
+
+    protected static void parseInputFile(final File inputFile,
```
1. Could use javadoc.
2. Does picard use any library that has a decent `Either` such that we could return an Either with the payload or the list of error messages?  That'd be nicer than taking mutable parameters.  If not could use `Tuple` from htsjdk to return two things
3. It's a bit weird to me that this returns/makes `BarcodeMetric` instances - are those always the thing thats needed?  Maybe we could define a little `BarcodeMetadata` class, return that, and have it have a `toBarcodeMetric()` method to build a metric object when it's needed?

jacarey replied:
For 3. I'm pretty sure we only populate the metrics map and that's all we do with the information coming back from the method. So if we did create a `BarcodeMetadata` object we then turn around and call `toBarcodeMetric()` on it and then nothing else, but I'll double check.

tfenne reviewed (commented):
I didn't look much at the tests.  One bigger thing is that I don't see collection of and then writing out of metrics happening anywhere except for in `ExtractIlluminaBarcodes`, but I think we still want the metrics when we run `IlluminaBasecallsToX` with on-the-fly demux.  We'll have to be careful too not to update metrics from multiple threads simultaneously, ugh.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,308 @@
+
+        // Prepopulate the lookup map with all perfect barcodes
+        for(BarcodeMetric metric : barcodeToMetrics.values()){
+            BarcodeExtractor.BarcodeMatch match = calculateBarcodeMatch(metric.barcodeBytes,
+                    perfectScores, true);
```
```suggestion
            BarcodeExtractor.BarcodeMatch match = calculateBarcodeMatch(metric.barcodeBytes, perfectScores, true);
```

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,308 @@
+     */
+    BarcodeMatch findBestBarcode(final byte[][] readSubsequences,
+                                 final byte[][] qualityScores,
+                                 final boolean isInlineMatching) {
```
From context I can't figure out what `isInlineMatching` does - could you add an `@param` for it please?

tfenne replied:
Ok, I figured it out (duh @tfenne ) but still think some parameter documentation on what it is and what effects it has would be good.

tfenne commented on src/main/java/picard/illumina/BarcodeExtractor.java:
```diff
@@ -0,0 +1,308 @@
+
+    BarcodeMatch calculateBarcodeMatch(final byte[][] readSubsequences,
+                                       final byte[][] qualityScores,
+                                       boolean isInlineMatching) {
```
```suggestion
                                       final boolean isInlineMatching) {
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -339,9 +344,45 @@ protected void setTileLimits(final Integer firstTile, final Integer tileLimit) {
         }
     }
 
+    protected String maybeDemultiplex(ClusterData cluster, Map<String, BarcodeMetric> metrics, BarcodeMetric noMatch) {
```
```suggestion
    /** If demultiplexing is to be performed, identify and return the barcode the cluster is assigned to. Otherwise returns null. */
    protected String maybeDemultiplex(ClusterData cluster, Map<String, BarcodeMetric> metrics, BarcodeMetric noMatch) {
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -339,9 +344,45 @@ protected void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+                    readSubsequences[i] = barcodeRead.getBases();
+                    qualityScores[i] = barcodeRead.getQualities();
+                }
+                BarcodeExtractor.BarcodeMatch match = barcodeExtractor.findBestBarcode(readSubsequences,
+                        qualityScores, true);
```
```suggestion
                BarcodeExtractor.BarcodeMatch match = barcodeExtractor.findBestBarcode(readSubsequences, qualityScores, true);
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -339,9 +344,45 @@ protected void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+
+                BarcodeExtractor.updateMetrics(match, cluster.isPf(), metrics, noMatch);
+
+                if(match.isMatched()) barcode = match.getBarcode();
```
```suggestion
                if (match.isMatched()) barcode = match.getBarcode();
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -339,9 +344,45 @@ protected void setTileLimits(final Integer firstTile, final Integer tileLimit) {
                 + " tasks were still running or queued and have been cancelled.");
     }
+
+    protected synchronized void updateMetrics(Map<String, BarcodeMetric> metrics, BarcodeMetric noMatch) {
```
```suggestion
    /** Updates the single/central demultiplexing metrics based on the metrics accumulated in another thread. */
    protected synchronized void updateMetrics(Map<String, BarcodeMetric> metrics, BarcodeMetric noMatch) {
```

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,199 @@
+    @Argument(doc = ReadStructure.PARAMETER_DOC, shortName = "RS")
+    public String READ_STRUCTURE;
+
+    @Argument(shortName = "GZIP", doc = "Compress output FASTQ files using gzip and append a .gz extension to the file names.")
+    public boolean COMPRESS_OUTPUTS = false;
```
Is this common to all programs?  I.e. when barcode-calling on the fly we don't write .barcode.txt files so maybe this only applies to ExtractIlluminaBarcodes?

jacarey replied:
This is whether or not we compress the fastqs as well.

tfenne commented on src/main/java/picard/illumina/ExtractBarcodesProgram.java:
```diff
@@ -0,0 +1,199 @@
+    private final NumberFormat tileNumberFormatter = NumberFormat.getNumberInstance();
+
+    /**
+     * The read structure of the actual Illumina Run, i.e. the readStructure of the input data
+     */
+    protected ReadStructure readStructure;
```
Should we maybe call it `inputReadStructure` or something similar then?

tfenne commented on src/main/java/picard/illumina/ExtractIlluminaBarcodes.java:
```diff
@@ -255,27 +171,31 @@ protected int doWork() {
         LOG.info("Processing with " + numProcessors + " PerTileBarcodeExtractor(s).");
         final ThreadPoolExecutor pool = new ThreadPoolExecutorWithExceptions(numProcessors);
 
+        // Create BarcodeMetric for counting reads that don't match any barcode
+        final String[] noMatchBarcode = new String[readStructure.sampleBarcodes.length()];
+        int index = 0;
+        for (final ReadDescriptor d : readStructure.descriptors) {
+            if (d.type == ReadType.Barcode) {
+                noMatchBarcode[index++] = StringUtil.repeatCharNTimes('N', d.length);
+            }
+        }
```
This is repeated in a handful of locations - perhaps extract to a static function somewhere?

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -75,6 +73,10 @@
             "<a href='http://broadinstitute.github.io/picard/command-line-overview.html#ExtractIlluminaBarcodes'>ExtractIlluminaBarcodes</a> " +
             "tool.</p>     " +
             "" +
+            "<p>Barcode matching can be done inline without requiring barcodes files generated by `ExtractIlluminaBarcode`. " +
+            "By setting MATCH_BARCODES_INLINE to true barcodes will be matched as they are parsed and converted. This" +
+            "does not require BARCODES_DIR.</p>     " +
+            "" +
```
Yay!  🎉

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -197,23 +186,32 @@
         final Set<String> barcodes = sampleBarcodeClusterWriterMap.keySet();
         basecallsConverter.processTilesAndWritePerSampleOutputs(barcodes);
+        // Finish metrics tallying.
+        if(METRICS_FILE != null && MATCH_BARCODES_INLINE) {
```
```suggestion
        if (METRICS_FILE != null && MATCH_BARCODES_INLINE) {
```

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -281,6 +275,12 @@ private void initialize() {
                 .withIgnoreUnexpectedBarcodes(IGNORE_UNEXPECTED_BARCODES)
                 .withBclQualityEvaluationStrategy(bclQualityEvaluationStrategy);
 
+        if(MATCH_BARCODES_INLINE && demultiplex){
```
```suggestion
        if (MATCH_BARCODES_INLINE && demultiplex) {
```

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -259,6 +252,10 @@
     protected int doWork() {
         initialize();
         basecallsConverter.processTilesAndWritePerSampleOutputs(barcodeSamWriterMap.keySet());
+        if(METRICS_FILE != null && MATCH_BARCODES_INLINE) {
```
```suggestion
        if (METRICS_FILE != null && MATCH_BARCODES_INLINE) {
```

tfenne replied:
![image](https://user-images.githubusercontent.com/1609210/109357292-37d65c00-783f-11eb-80c5-826d0acc8c64.png)

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToSam.java:
```diff
@@ -301,6 +298,12 @@ private void initialize() {
                 .withIgnoreUnexpectedBarcodes(IGNORE_UNEXPECTED_BARCODES)
                 .withBclQualityEvaluationStrategy(bclQualityEvaluationStrategy);
 
+        if(MATCH_BARCODES_INLINE && demultiplex){
```
```suggestion
        if (MATCH_BARCODES_INLINE && demultiplex) {
```

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,119 @@
+            }).collect(Collectors.toList());
+
+            if (readStructure.sampleBarcodes.length() != validBarcodeColumns.size()) {
+                messages.add("Expected " + readStructure.sampleBarcodes.length() + " valid barcode columns, but only found " +
```
```suggestion
                messages.add("Expected " + readStructure.sampleBarcodes.length() + " valid barcode columns, but found " +
```

tfenne replied:
Removed the `only` since in theory you could have more too

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,119 @@
+                        String.join(",", validBarcodeColumns));
+            }
+
+            Matcher matcher = Pattern.compile("^(.*)_\\d").matcher(validBarcodeColumns.get(0));
```
Do you care which one you get?  Might it matter if the user has put `BARCODE_2` before `BARCODE`?  If so might be good to sort `validBarcodeColumns` first by length and then by `String.compare()`?

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,119 @@
+
+            Matcher matcher = Pattern.compile("^(.*)_\\d").matcher(validBarcodeColumns.get(0));
+
+            final String sequenceColumn;
+            boolean hasMultipleNumberedBarcodeColumns = matcher.matches();
+            if (hasMultipleNumberedBarcodeColumns) {
+                sequenceColumn = matcher.group(1);
+            } else {
+                sequenceColumn = validBarcodeColumns.get(0);
+            }
```
I find this a bit hard to follow - perhaps add javadoc with what the intent is?

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,119 @@
+        List<String> messages = new ArrayList<>();
+        Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
+        try (final TabbedTextFileWithHeaderParser barcodesParser = new TabbedTextFileWithHeaderParser(inputFile)) {
+            List<String> validBarcodeColumns = barcodesParser.columnLabels().stream().filter(name -> {
+                boolean isValidPrefix = false;
+                for (String columnPrefix : BARCODE_PREFIXES) {
+                    isValidPrefix |= name.toUpperCase().startsWith(columnPrefix.toUpperCase());
+                }
+                return isValidPrefix;
+            }).collect(Collectors.toList());
```
I think I would a) extract some of this logic into a function called `getNameOfFirstBarcodeColumn()` that looks something like:

```java
final String getNameOfFirstBarcodeColumn(final Set<String> cols) {
   for (final String prefix : BARCODE_PREFIXES) {
        if (cols.contains(prefix)) return prefix;
        if (cols.contains(prefix + "_1") return prefix + "_1;
    }

    throw new IllegalStateException("Couldn't find column containing first barcode sequence. Expected ...");
}
```

then b) in the loop below use the returned value if `barcodeNum = 0` else use the prefix plus the `_barcodeNum`

jacarey replied:
Yeah its a bit more annoying than that since a single barcode is just `BARCODE` and multiple would be `BARCODE_1` `BARCODE_2` so depending on how many valid columns you have you'd either be appending the numbered suffix or not. 

I'll try to rework this though.

Also we can't really do a contains and keep case insensitivity without some rearranging.

tfenne commented on src/main/java/picard/illumina/SampleInputParameters.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Parses any one of the following types of files:
+     *
+     * ExtractIlluminaBarcodes      BARCODE_FILE
+     * IlluminaBasecallsToFastq     MULTIPLEX_PARAMS
+     * IlluminaBasecallsToSam       LIBRARY_PARAMS
```
So happy to see this!  One file format for all of demultiplexing!

tfenne commented on src/main/java/picard/illumina/SortedBasecallsConverter.java:
```diff
@@ -140,10 +142,20 @@ public void run() {
         TileProcessor(final int tileNum, final Set<String> barcodes) {
             this.tileNum = tileNum;
             this.barcodeToRecordCollection = new HashMap<>(barcodes.size(), 1.0f);
+            if(barcodeExtractor != null) {
```
```suggestion
            if (barcodeExtractor != null) {
```

tfenne reviewed (commented):
A lot of mostly minor comments and suggestions.  Thanks @jacarey !

Commit:
Add on the fly demux and speed up barcode matching

Commit:
Increment atomic int prior to job submission.

Commit:
Only read the cbcl header once the tile is about to be cached and then close the streams. (#1686)

* Only read header once the tile is about to be cached and then close the cbcl streams.
* Fix header read for check illumina dir.
* Close provider after we read all clusters.

Commit:
Merge branch 'master' into tf_speedup_extract_illumina_barcodes

Commit:
[bugfix] Pass MAX_RECORDS_IN_RAM to BasecallsConverter (#1687)

Commit:
Jc fix small data race condition (#1690)

* [bugfix] Pass MAX_RECORDS_IN_RAM to BasecallsConverter
* Use a different thread pool executor for each tile to ensure no race conditions can occur

Commit:
Jc test unsorted threading (#14)

* Add basic unsorted threading

Commit:
bugfix and nio access in Fingerprinting code (no need to panic) (#1694)

* - Convert CalculateFingerprintMetrics, ExtractFingerprint, and ExtractContamination to use NIO and thus be able to to their work on a non-localized input file.
- Fixed a bug in the AD output of extractFingerprint and CalculateContamination (Ref and Alt were swapped...)
- Renamed PL variable to be GL since that's what is actually was.

* -removed unneeded mathUtil method.

Commit:
fix: underflow in HaplotypeProbabilitiesFromContaminatorSequence (#1697)

* fix: underflow in HaplotypeProbabilitiesFromContaminatorSequence, which affects both IdentifyContamination and ExtractFingerprint in deep coverage. This was leading to non-informative sites in RNA and similar sequencing data.

Commit:
Rebase and fix for UnsortedBasecallsConverter + IGNORE_UNEXPECTED_BARCODES = true

mollysacks committed:
Fix bug in CollectSamErrorMetrics (#1618)

* first

* place-holder

* testing

* testing

* delete-logfiles

* rm-ds-store

* Update CollectSamErrorMetrics.java

* add-test-case

Co-authored-by: Molly Sacks <msacks@broadinstitute.org>

mollysacks committed:
HaplotypeMap bug fix (#1702)

* use-basesEqual

* testing

Co-authored-by: Molly Sacks <msacks@broadinstitute.org>

gbggrant committed:
Fix typo in usage (#1703)

Commit:
Review changes.

Commit:
Merge branch 'master' into tf_speedup_extract_illumina_barcodes
